lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -592,27 +655,39 @@ monitorStakePools tr gp nl DBLayer{..} =
 monitorMetadata
     :: Tracer IO StakePoolLog
     -> GenesisParameters
-    -> (   PoolId
-        -> StakePoolMetadataUrl
-        -> StakePoolMetadataHash
-        -> IO (Maybe StakePoolMetadata)
-       )
     -> DBLayer IO
+    -> MVar Bool
     -> IO ()
-monitorMetadata tr gp fetchMetadata DBLayer{..} = forever $ do
-    refs <- atomically (unfetchedPoolMetadataRefs 100)
-    successes <- fmap catMaybes $ forM refs $ \(pid, url, hash) -> do
-        fetchMetadata pid url hash >>= \case
-            Nothing -> Nothing <$ do
-                atomically $ putFetchAttempt (url, hash)
-
-            Just meta -> Just hash <$ do
-                atomically $ putPoolMetadata hash meta
-
-    when (null refs || null successes) $ do
-        traceWith tr $ MsgFetchTakeBreak blockFrequency
-        threadDelay blockFrequency
+monitorMetadata tr gp DBLayer{..} stop = do
+    settings <- atomically readSettings
+    loop settings
   where
+    loop settings = do
+        (refs, successes) <- case poolMetadataSource settings of
+            FetchNone -> pure ([], [])
+            FetchDirect -> (fetcher [identityUrlBuilder]) >>= fetchThem
+            FetchSMASH (unSmashServer -> uri) -> (fetcher [registryUrlBuilder uri]) >>= fetchThem
+
+        when (null refs || null successes) $ do
+            traceWith tr $ MsgFetchTakeBreak blockFrequency
+            threadDelay blockFrequency
+        readMVar stop >>= \case
```

Comment:
I agree with @KtorZ - because this thread is a polling loop, it doesn't need to stop. Perhaps we confused ourselves initially thinking that some thread needs to be restarted for changed settings to take effect. No, that's not the case - the settings API request handler can wipe the metadata - next time this loop goes around, it can read settings from the DB and proceed to fetch any missing metadata.

Reply:
We're running in circles. I already implemented that approach and we discussed it and that I don't like to poll the database every loop for the settings (it's a waste of resources).

The other two approaches are:

1. Simple **restart-logic**, which is very general and can be applied *regardless* of the internals of the data (this is good). Even places that have no access or knowledge of the data can in theory restart the thread.
2. Data synchronization logic... the `putSettings` endpoint will put the new Settings value additionally into an MVar... I dislike this approach for having the mutable Settings in two places and being data specific.

Reply:
2. is much simpler and doesn't require extra synchronization and restart logic. It doesn't actually create two points of mutation for the settings, but it does create a synchronized view of them between the thread modifying the database, and the thread which actually needs to access the settings.

Reply:
> I don't like to poll the database every loop for the settings (it's a waste of resources).

It's a waste of _a very small amount of_ resources.

Reply:
> is much simpler and doesn't require extra synchronization and restart logic. It doesn't actually create two points of mutation for the settings, but it does create a synchronized view of them between the thread modifying the database, and the thread which actually needs to access the settings.

I think for the problem at hand, all 3 implementations are viable and the difference between them isn't very big. However, I've spent some time implementing the thread restart logic and I quite like it. If I go for 2. then I will have implemented every single one of them in the course of this PR :smile: 

Thread restart logic has also the advantage that we could move more logic into the thread start part:

```hs
monitorMetadata tr gp DBLayer{..} stop = do
    -- thread start logic
    settings <- atomically readSettings

   -- thread loop
    loop settings
```

So we could even move the dropping of the table into this part, instead of convoluting `pushSettings` with that logic and not worry about the performance of the start phase, since it's not part of the loop.